workflows/dnaseq/dnaseq-core.wdl

```diff
@@ -0,0 +1,160 @@
+        aligner: {
+            description: "BWA aligner to use",
+            choices: ["mem", "aln"]
+        }
+        validate_input: "Ensure input BAM is well-formed before beginning harmonization?"
+        use_all_cores: "Use all cores? Recommended for cloud environments."
+    }
+    input {
+        Array[File] read_one_fastqs_gz
+        Array[File] read_two_fastqs_gz
+        File bwa_db
+        Int reads_per_file = 10000000
+        Array[ReadGroup] read_groups
+        String prefix
+        String aligner = "mem"
+        Boolean use_all_cores = false
+    }
+
+    call parse_input { input:
+        aligner=aligner
+    }
+
+    scatter (rg in read_groups) {
+        call read_group.ReadGroup_to_string { input: read_group=rg }
+    }
+
+    Array[String] read_groups_bwa = prefix("@RG\t", ReadGroup_to_string.stringified_read_group)
+
+    scatter (tuple in zip(
+        zip(read_one_fastqs_gz, read_two_fastqs_gz),
+        read_groups_bwa
+    )) {
+        call util.split_fastq as read_ones { input:
+            fastq = tuple.left.left,
+            reads_per_file = reads_per_file
+        }
+
+        call util.split_fastq as read_twos { input:
+            fastq = select_first([tuple.left.right, "undefined"]),
+            reads_per_file = reads_per_file
+        }
+
+        scatter (t in zip(read_ones.fastqs, read_twos.fastqs)) {
+            if (aligner == "mem") {
+                call bwa.bwa_mem { input:
+                    read_one_fastq_gz=t.left,
+                    read_two_fastq_gz=select_first([t.right, "undefined"]),
+                    bwa_db_tar_gz=bwa_db,
+                    # find tab literals, replace with '\\t' (which must be written as '\\\\t')
+                    # '\\t' is subbed into command blocks as '\t'
+                    read_group=sub(tuple.right, "\t", "\\\\t"),
+                    use_all_cores=use_all_cores,
+                }
+            }
+            if ( aligner == "aln") {
+                call bwa.bwa_aln_pe { input:
+                    read_one_fastq_gz=t.left,
+                    read_two_fastq_gz=select_first([t.right, "undefined"]),
+                    bwa_db_tar_gz=bwa_db,
+                    # find tab literals, replace with '\\t' (which must be written as '\\\\t')
+                    # '\\t' is subbed into command blocks as '\t'
+                    read_group=sub(tuple.right, "\t", "\\\\t"),
+                    use_all_cores=use_all_cores,
+                }
+            }
+            call picard.sort as sort { input:
+                 bam=select_first([bwa_mem.bam, bwa_aln_pe.bam])
+            }
+        }
+
+        call samtools_merge_wf.samtools_merge as inner_merge { input:
+            bams = sort.sorted_bam,
+            prefix = basename(tuple.left.left, ".fastq.gz"),
+            use_all_cores = use_all_cores,
+        }
+    }
+    call samtools_merge_wf.samtools_merge as rg_merge { input:
```

Comment:
Can we use `flatten()` to only call `samtools_merge_wf.samtools_merge` once? That would make this workflow easier to follow IMO.

Reply:
Maybe? Based on the comment above, I already need to look at this again. I want to get `samtools merge` to correctly collapse `@RG` and `@PG` records. I'm not certain that the `@PG` records will collapse properly since the paths will be different.

Reply:
That worked, so I've made the change.